### DISCUSSION TITLE
shouldn't IVF_PQ nbits param match the number specified in nlist param?

### DISCUSSION THREAD

**Discussion by MazenH24** [milvus-io/milvus#21045] [Q&A and General discussion]
I noticed that I can define an IVF_PQ with the following params
`index_params = {
    "index_type": "IVF_PQ",
    "metric_type": "L2",
    "params": {
        "nlist": 500,
        "m": 32,
        "nbits": 8
    }
}`
using this set of params doesn't generate any error and the index is created.
isn't it wrong to specify nbits as 8 here? 
since 8 bits can represent only 2^8 = 256, but the number of clusters is 500 which needs at least 9 bits in order not to have a bit overflow.
e.g. PQ code for segment i is 257 using 8 bit encodings the identifier for the the segment will be stored as 1, (257) = (1 0000 0001).
Am I missing something here or an error must be raised when specifying such a set of params?

**Comment by yhmo** [comment]
Please refer this document: https://milvus.io/docs/index.md





Parameter | Description | Range
-- | -- | --
nlist | Number of cluster units | [1, 65536]
m | Number of factors of product quantization | dim ≡ 0 (mod m)
nbits | [Optional] Number of bits in which each low-dimensional vector is stored. | [1, 16] (8 by default)

**Reply by yhmo** [reply]
IVF_PQ first creates `nlist` clusters, and then applies PQ to each cluster. So, the `nbit` value has nothing to do with `nlist` value.

**Comment by MazenH24** [comment]
Thank you!
I read the docs, but I was confused with another implementation of PQ
So if I understood it right, we cluster the initial vectors into nlist clusters, then for each cluster we split the vectors into m subspaces, and for each subspace we run clustering again to cluster each segment set into 2^nbit clusters, is that right?

**Reply by yhmo** [reply]
> Thank you! I read the docs, but I was confused with another implementation of PQ So if I understood it right, we cluster the initial vectors into nlist clusters, then for each cluster we split the vectors into m subspaces, and for each subspace we run clustering again to cluster each segment set into 2^nbit clusters, is that right?

right
